rfcs/regex_feature.md

```diff
@@ -0,0 +1,91 @@
+# Regex feature
+
+A feature based on regex pattern matching that MindMeld's NER models can use to detect custom entities. 
+
+- Developers expect NER models to correctly infer that the query `AVET128-321` is a `pid` entity 
+without providing surrounding sentence context to the classifier. This is more relevant in services that
+just use MindMeld for NLP while doing their own dialog management. 
+- Developers also want granular distinctions between entities that occur in the same context, 
+for example, `look up {SKU120|sku}` and `look up {123-ABC-12|pid}`. Currently, we ask developers to differ
+these types of distinction to the business logic in the dialogue manager.
+- DialogFlow, NLTK and other similar platforms have the regex matching feature. Moreover, we already have an 
+orthographic matching feature in MindMeld. This feature would augment that.
+
+
+## Usage
+
+The developer makes a new file called `regex_matches.txt` for the entity type in the entities folder that needs 
+these regex matches like this following:
+
+```text
+entities
+├── account_type
+│   ├── gazetteer.txt
+│   ├── mapping.json
+│   └── regex_matches.txt
+├── autopay_status
+│   └── mapping.json
+└── credit_amount
+    └── mapping.json
+```
+
+Inside the `regex_matches.txt` file, the developer lists regex patterns compliant with python syntax that need 
+to be matched for this custom entity. 
+
+```text
+\w{4}[0-9]{3}\-[0-9]{3}
+[0-9]{3}\-[0-9]{3}
+```
+
+In the above syntax, the `account_type` entity would match words like `ABCD-123-281` (first pattern) 
+and `123-456` (second pattern).
+
+Finally, the developer would have the ability to turn on/off the regex feature through the model config in 
+config.py using the key,value `regex-seq: True`. The feature will be off by default.
+
+```python
+ENTITY_RECOGNIZER_CONFIG = {
+    "features": {
+        "bag-of-words-seq": {
+            "ngram_lengths_to_start_positions": {
+                1: [-2, -1, 0, 1, 2],
+                2: [-2, -1, 0, 1],
+            }
+        },
+        "sys-candidates-seq": {"start_positions": [-1, 0, 1]},
+        "regex-seq": True
+    }
+}
+```
+
+If one is using the `view_extracted_features` method of the entity classifier to inspect the features of a query,
+the following result would be shown:
+
+```python
+er.view_extracted_features("look up 123-456")
+```
+
+```text
+[{...}, {...}, {'regex_match|type:account_type:1', ...}]
+```
+
+## Problems with this approach
+
+- The processing time during inference will increase significantly since each input token has to be matched against 
+each regex pattern for every entity type.
+- The training time will also increase for the similar reason
```

Comment:
What if the user just wants regex matching for an entity (which seems like the most common usecase to me)? If my understanding is right, in the current setup, the user would end up creating a model that only uses regex matching as a feature. That seems redundant? Maybe the user can add it either as a feature or a separate key in the entity config. If it's the latter, we just use regex matching as the ER algo.

Reply:
`What if the user just wants regex matching for an entity (which seems like the most common usecase to me)`: I think this is a valid use-case, but the solution proposed `user can add it either as a feature or a separate key in the entity config` will lock the platform into this singular use-case.

I think users would like to complement existing signals, like bag-of-words, gazetteers etc with the regex matching feature to capture a regular pattern of queries that are not being recognized with the former approaches (without having to add every combination to gaz). 

Maybe the architecture could be smart in a way where if the only feature is the regex feature, we could short-circuit the model building and just use that to label queries but use the models if not. One issue with approach is there could be collisions between the regex tagger and the normal entity tagger.